# E-Commerce SQL Analytics Project

In [3]:
from sqlalchemy import create_engine
import pandas as pd

engine = create_engine("mysql+pymysql://root:akhil123@localhost:3306/ecommerce_project")


## Setup: Load SQL Extension and Connect to MySQL


In [2]:
%load_ext sql


In [3]:
%%sql
mysql+pymysql://root:akhil123@localhost:3306


## Raw Data Table

In [14]:
%%sql
use ecommerce_project;
CREATE TABLE raw_data (
    Order_ID VARCHAR(30),
    Customer_ID VARCHAR(30),
    Date VARCHAR(30),
    Age INT,
    Gender VARCHAR(10),
    City VARCHAR(50),
    Product_Category VARCHAR(50),
    Unit_Price DECIMAL(10,2),
    Quantity INT,
    Discount_Amount DECIMAL(10,2),
    Total_Amount DECIMAL(10,2),
    Payment_Method VARCHAR(30),
    Device_Type VARCHAR(20),
    Session_Duration_Minutes INT,
    Pages_Viewed INT,
    Is_Returning_Customer BOOLEAN,
    Delivery_Time_Days INT,
    Customer_Rating INT
);


 * mysql+pymysql://root:***@localhost:3306
0 rows affected.
0 rows affected.


[]

## Loading the Raw CSV Data and Push to MySQL
This step reads the e-commerce dataset into a Pandas DataFrame and inserts all records into the `raw_data` table inside the MySQL database.


In [15]:
df = pd.read_csv("D:\\PROJECT\\Ecommerce-SQL-Analysis\\Data\\ecommerce_customer_behavior_dataset_v2.csv")
df.to_sql("raw_data", engine, if_exists="append", index=False, method="multi")


17049

## Create `customers` Table

In [16]:
%%sql
CREATE TABLE customers (
    customer_id VARCHAR(30) PRIMARY KEY,
    age INT,
    gender VARCHAR(10),
    city VARCHAR(50),
    is_returning_customer BOOLEAN
);


 * mysql+pymysql://root:***@localhost:3306
0 rows affected.


[]

## Insert Distinct Customer Records into `customers`

In [17]:
%%sql
INSERT INTO customers (customer_id, age, gender, city, is_returning_customer)
SELECT
    Customer_ID,
    MIN(Age) AS age,
    MIN(Gender) AS gender,
    MIN(City) AS city,
    MAX(Is_Returning_Customer) AS is_returning_customer
FROM raw_data
GROUP BY Customer_ID;


 * mysql+pymysql://root:***@localhost:3306
5000 rows affected.


[]

## Create `products` Table

In [18]:
%%sql
CREATE TABLE IF NOT EXISTS products (
    product_id INT AUTO_INCREMENT PRIMARY KEY,
    product_category VARCHAR(50),
    unit_price DECIMAL(10,2)
);


 * mysql+pymysql://root:***@localhost:3306
0 rows affected.


[]

## Insert Product Records into `products`

In [19]:
%%sql
INSERT INTO products (product_category, unit_price)
SELECT DISTINCT
    Product_Category,
    Unit_Price
FROM raw_data;


 * mysql+pymysql://root:***@localhost:3306
16402 rows affected.


[]

## Create `orders` Table

In [20]:
%%sql
CREATE TABLE IF NOT EXISTS orders (
    order_id VARCHAR(30) PRIMARY KEY,
    customer_id VARCHAR(30),
    product_id INT,
    date DATE,
    quantity INT,
    discount_amount DECIMAL(10,2),
    total_amount DECIMAL(10,2),
    payment_method VARCHAR(30),
    FOREIGN KEY (customer_id) REFERENCES customers(customer_id),
    FOREIGN KEY (product_id) REFERENCES products(product_id)
);


 * mysql+pymysql://root:***@localhost:3306
0 rows affected.


[]

## Insert Order Records into `orders`

In [21]:
%%sql
INSERT INTO orders (
    order_id, customer_id, product_id, date,
    quantity, discount_amount, total_amount, payment_method
)
SELECT
    r.Order_ID,
    r.Customer_ID,
    p.product_id,
    STR_TO_DATE(r.Date, '%Y-%m-%d'),
    r.Quantity,
    r.Discount_Amount,
    r.Total_Amount,
    r.Payment_Method
FROM raw_data r
JOIN products p
ON r.Product_Category = p.product_category
AND r.Unit_Price = p.unit_price;


 * mysql+pymysql://root:***@localhost:3306
17049 rows affected.


[]

## Create `customer_behavior` Table

In [22]:
%%sql
CREATE TABLE IF NOT EXISTS customer_behavior (
    order_id VARCHAR(30),
    device_type VARCHAR(20),
    session_duration_minutes INT,
    pages_viewed INT,
    FOREIGN KEY (order_id) REFERENCES orders(order_id)
);


 * mysql+pymysql://root:***@localhost:3306
0 rows affected.


[]

## Insert Behavior Data into `customer_behavior`

In [23]:
%%sql
INSERT INTO customer_behavior (
    order_id, device_type, session_duration_minutes, pages_viewed
)
SELECT
    Order_ID,
    Device_Type,
    Session_Duration_Minutes,
    Pages_Viewed
FROM raw_data;


 * mysql+pymysql://root:***@localhost:3306
17049 rows affected.


[]

##  Create `delivery_reviews` Table

In [24]:
%%sql
CREATE TABLE IF NOT EXISTS delivery_reviews (
    order_id VARCHAR(30),
    delivery_time_days INT,
    customer_rating INT,
    FOREIGN KEY (order_id) REFERENCES orders(order_id)
);


 * mysql+pymysql://root:***@localhost:3306
0 rows affected.


[]

## Insert Delivery Review Data into `delivery_reviews`

In [25]:
%%sql
INSERT INTO delivery_reviews (
    order_id, delivery_time_days, customer_rating
)
SELECT
    Order_ID,
    Delivery_Time_Days,
    Customer_Rating
FROM raw_data;


 * mysql+pymysql://root:***@localhost:3306
17049 rows affected.


[]

# 📊Exploratory Data Analysis (SQL)

<br>

## 1. Total Revenue Analysis
**Result:** $21,779,052.59

### 💡 Key Insights:
- **Scale Assessment:** With $21.78M in revenue across 15 months, the average monthly revenue is ~$1.45M
- **Baseline Established:** This provides our benchmark for measuring growth and setting future targets
- **Market Position:** For a mid-sized e-commerce operation, this indicates healthy transaction volume

In [31]:
%%sql
SELECT SUM(total_amount) AS total_revenue FROM orders;


 * mysql+pymysql://root:***@localhost:3306
1 rows affected.


[(Decimal('21779052.59'),)]

## 2. Monthly Revenue Trend
**Peak:** December 2023 ($1,586,214) | **Low:** January 2023 ($1,311,603)

### 💡 Key Insights:
- **Seasonality Detected:** 
  - Q4 2023 shows strong performance (Oct-Dec avg: $1.52M)
  - Q1 2024 declining trend (Jan-Mar avg: $1.36M) - **10.5% drop**
- **Holiday Peak:** December spike (+9.7% vs November) confirms holiday shopping impact
- **Concerning Trend:** March 2024 ($1.27M) is the lowest month in the dataset - **19.8% below December**

### 📊 Pattern Analysis:
```
Strong Months: July ($1.56M), December ($1.59M)
Weak Months: January 2023 ($1.31M), March 2024 ($1.27M)
Grow

In [32]:
%%sql
SELECT 
    DATE_FORMAT(date, '%Y-%m') AS month,
    SUM(total_amount) AS revenue
FROM orders
GROUP BY month
ORDER BY month;


 * mysql+pymysql://root:***@localhost:3306
15 rows affected.


[('2023-01', Decimal('1311603.21')),
 ('2023-02', Decimal('1409777.29')),
 ('2023-03', Decimal('1458005.58')),
 ('2023-04', Decimal('1499648.03')),
 ('2023-05', Decimal('1483511.61')),
 ('2023-06', Decimal('1380984.94')),
 ('2023-07', Decimal('1560046.14')),
 ('2023-08', Decimal('1510999.75')),
 ('2023-09', Decimal('1502164.37')),
 ('2023-10', Decimal('1536847.33')),
 ('2023-11', Decimal('1446439.79')),
 ('2023-12', Decimal('1586213.77')),
 ('2024-01', Decimal('1490824.78')),
 ('2024-02', Decimal('1329884.01')),
 ('2024-03', Decimal('1272101.99'))]

<br>

## 3. Top 10 Cities by Revenue

**Winner:** Istanbul ($5.65M - 26% of total revenue)

### 💡 Key Insights:
- **Extreme Concentration:** Top 3 cities (Istanbul, Ankara, Izmir) = **52% of total revenue**
- **Opportunity Gap:** Istanbul generates 85% MORE than Ankara (#2)
- **Long Tail:** 7 cities below $2M each suggests untapped markets

### 📊 Geographic Analysis:
```
Tier 1: Istanbul ($5.65M) - 26%
Tier 2: Ankara ($3.05M), Izmir ($2.65M) - 26% combined
Tier 3: Bursa, Antalya, Gaziantep - 25% combined
Tier 4: Others - 23%

In [5]:
%%sql
use ecommerce_project;
SELECT 
    c.city,
    SUM(o.total_amount) AS total_revenue
FROM orders o
JOIN customers c ON o.customer_id = c.customer_id
GROUP BY c.city
ORDER BY total_revenue DESC
LIMIT 10;


 * mysql+pymysql://root:***@localhost:3306
0 rows affected.
10 rows affected.


[('Istanbul', Decimal('5646595.78')),
 ('Ankara', Decimal('3052443.10')),
 ('Izmir', Decimal('2651743.92')),
 ('Bursa', Decimal('2317028.10')),
 ('Antalya', Decimal('1661594.15')),
 ('Gaziantep', Decimal('1573422.25')),
 ('Adana', Decimal('1512320.19')),
 ('Konya', Decimal('1363152.80')),
 ('Kayseri', Decimal('1091658.53')),
 ('Eskisehir', Decimal('909093.77'))]

<br>

## 4. Best-Selling Product Categories

**Champion:** Electronics ($10.48M - 48% of revenue)

### 💡 Key Insights:
- **Dominance:** Electronics alone = nearly HALF of all revenue
- **High-Value, Low-Volume:** Electronics: 6,277 units but $10.48M (avg $1,670/unit)
- **High-Volume, Low-Value:** Books: 6,727 units but only $360K (avg $54/unit)
- **Balanced Performer:** Home & Garden: Good revenue ($4.02M) with decent volume (6,155)

In [34]:
%%sql
SELECT 
    p.product_category,
    SUM(o.total_amount) AS revenue,
    SUM(o.quantity) AS units_sold
FROM orders o
JOIN products p ON o.product_id = p.product_id
GROUP BY p.product_category
ORDER BY revenue DESC;


 * mysql+pymysql://root:***@localhost:3306
8 rows affected.


[('Electronics', Decimal('10481897.65'), Decimal('6277')),
 ('Home & Garden', Decimal('4023903.94'), Decimal('6155')),
 ('Sports', Decimal('3205086.99'), Decimal('6856')),
 ('Fashion', Decimal('1577035.70'), Decimal('6078')),
 ('Toys', Decimal('1014237.53'), Decimal('6303')),
 ('Beauty', Decimal('694437.02'), Decimal('6564')),
 ('Food', Decimal('422054.65'), Decimal('6381')),
 ('Books', Decimal('360399.11'), Decimal('6727'))]

<br>

## 5. Device-Based Customer Spending

**Highest AOV:** Desktop ($1,311) | Mobile ($1,260) | Tablet ($1,258)

### 💡 Key Insights:
- **Desktop Premium:** Desktop users spend 4% more per order
- **Minimal Difference:** Only $53 separates highest and lowest (very balanced)
- **Mobile Volume:** Need to cross-reference with traffic data (likely highest volume)

### 📊 Strategic Implications:
- Desktop users may be:
  - B2B buyers (purchasing for offices)
  - More deliberate shoppers (researching before buying)
  - Buying higher-value electronics
  
- Mobile users:
  - Impulse purchases (lower consideration)
  - Convenience-driven (quick checkouts)


In [35]:
%%sql
SELECT 
    b.device_type,
    AVG(o.total_amount) AS avg_spend
FROM customer_behavior b
JOIN orders o ON b.order_id = o.order_id
GROUP BY b.device_type;


 * mysql+pymysql://root:***@localhost:3306
3 rows affected.


[('Mobile', Decimal('1260.472104')),
 ('Desktop', Decimal('1310.767677')),
 ('Tablet', Decimal('1257.634094'))]

<br>

## 6. Customer Lifetime Value (Top 10)

**Top Customer:** CUST_01573 ($50,628 from just 2 orders!)

### 💡 Key Insights:
- **Whale Customer:** Top customer spent $50K - likely B2B or reseller
- **High-Value Segment:** Top 10 customers = $400K+ (1.8% of total revenue from 0.2% of customers)
- **Order Frequency Varies:** 
  - Some VIPs order 10 times (steady engagement)
  - Others order 2-3 times (large bulk purchases)

### 📊 VIP Customer Profile:
```
Average CLV (Top 10): $40,000
Average Orders (Top 10): 7.6 orders
Average Order Value: $5,263

Comparison to Overall:
- Overall AOV: $1,277
- VIP AOV: 4.1x higher
```


In [39]:
%%sql
SELECT 
    customer_id,
    SUM(total_amount) AS lifetime_value,
    COUNT(order_id) AS total_orders,
    AVG(total_amount) AS avg_order_value
FROM orders
GROUP BY customer_id
ORDER BY lifetime_value DESC
limit 10;


 * mysql+pymysql://root:***@localhost:3306
10 rows affected.


[('CUST_01573', Decimal('50628.15'), 2, Decimal('25314.075000')),
 ('CUST_00197', Decimal('47324.23'), 10, Decimal('4732.423000')),
 ('CUST_03795', Decimal('42480.02'), 3, Decimal('14160.006667')),
 ('CUST_00154', Decimal('41613.83'), 10, Decimal('4161.383000')),
 ('CUST_04219', Decimal('40608.39'), 8, Decimal('5076.048750')),
 ('CUST_03149', Decimal('39834.61'), 6, Decimal('6639.101667')),
 ('CUST_03923', Decimal('38610.82'), 10, Decimal('3861.082000')),
 ('CUST_01164', Decimal('36842.41'), 9, Decimal('4093.601111')),
 ('CUST_01816', Decimal('36557.27'), 9, Decimal('4061.918889')),
 ('CUST_04246', Decimal('36295.30'), 9, Decimal('4032.811111'))]

<br>

## 7. Average Order Value Per Customer

**Result:** $1,262.89

### 💡 Key Insights:
- **Baseline Metric:** Average customer generates $1,263 per order
- **Comparison:**
  - VIP customers (top 10): $5,263 AOV (4.2x higher)
  - Regular customers: ~$1,200 AOV
- **Growth Potential:** Increasing AOV by just 10% = +$2.18M annual revenue


In [43]:
%%sql
WITH cust_avg AS (
  SELECT customer_id, AVG(total_amount) AS avg_order_value
  FROM orders
  GROUP BY customer_id
)
SELECT AVG(avg_order_value) AS avg_order_value_per_customer
FROM cust_avg;


 * mysql+pymysql://root:***@localhost:3306
1 rows affected.


[(Decimal('1262.8864468778'),)]

<br>


## 8. Average Age of Customers by City

**Key Finding:** Average age is remarkably consistent (34.4 - 35.5 years)

### 💡 Key Insights:
- **Demographic Uniformity:** All major cities cluster around 35 years old (±1 year variance)
- **Target Demographic Clear:** Core customer = 30-40 year old professionals
- **Istanbul Dominance Confirmed:** 1,284 customers (25.7% of customer base)
- **No Geographic Age Bias:** Konya (34.4) vs Antalya (35.5) - minimal difference

### 📊 Customer Distribution Analysis:
```
Large Markets (1000+ customers):
- Istanbul: 1,284 customers, avg age 35.1

Medium Markets (500-750 customers):
- Ankara: 735 customers, avg age 34.9
- Izmir: 600 customers, avg age 35.4

Small Markets (<500 customers):
- Remaining 7 cities: 3,381 customers combined
`

In [45]:
%%sql
SELECT
  city,
  AVG(age) AS avg_age,
  COUNT(customer_id) AS num_customers
FROM customers
GROUP BY city
ORDER BY num_customers DESC;


 * mysql+pymysql://root:***@localhost:3306
10 rows affected.


[('Istanbul', Decimal('35.0607'), 1284),
 ('Ankara', Decimal('34.9483'), 735),
 ('Izmir', Decimal('35.4100'), 600),
 ('Bursa', Decimal('34.9879'), 496),
 ('Adana', Decimal('34.5503'), 378),
 ('Antalya', Decimal('35.4679'), 374),
 ('Gaziantep', Decimal('35.4585'), 349),
 ('Konya', Decimal('34.4479'), 317),
 ('Kayseri', Decimal('34.8872'), 257),
 ('Eskisehir', Decimal('34.6286'), 210)]

<br>

## 9. Returning vs New Customers

**Critical Finding:** 93% are returning customers!

### 💡 Key Insights:
- **Retention Success:** 4,651 returning customers vs 349 new (93% vs 7%)
- **Loyalty Strength:** Returning customers average 3.6 orders each (16,700 orders / 4,651 customers)
- **New Customer Challenge:** Only 349 new customers acquired in 15 months = **23 new customers/month**

### 📊 Alarming Math:
```
Customer Acquisition:
- New customers: 349 in 15 months = 23/month
- If customer churn = 5% monthly: Losing 233 customers/month
- Net customer loss: -210 customers/month
- THIS IS A CRISIS: Business is shrinking!
```


In [ ]:
%%sql
SELECT
  c.is_returning_customer,
  COUNT(DISTINCT o.customer_id) AS num_customers,
  COUNT(o.order_id) AS num_orders,
  SUM(o.total_amount) AS revenue
FROM orders o                   #0-new customers, 1-old customers
JOIN customers c ON o.customer_id = c.customer_id
GROUP BY c.is_returning_customer;


 * mysql+pymysql://root:***@localhost:3306
2 rows affected.


[(0, 349, 349, Decimal('400770.69')), (1, 4651, 16700, Decimal('21378281.90'))]

<br>


## 10. Most Frequently Sold Products

**Top Product:** Product ID 712 (Books) - 17 units

### 💡 Key Insights:
- **High-Frequency, Low-Value:** Top products are Books, Food, Beauty (all low-price)
- **Volume ≠ Revenue:** Product 712 sold 17 units but only $396 revenue ($23/unit)
- **Missing High-Value Items:** No electronics in top 20 despite being 48% of revenue

### 📊 Business Interpretation:
```
High Frequency Products:
- Books: Sold often, low revenue
- Food: Repeat purchases, subscription potential
- Beauty: Consumables, need replenishment

High Value Products (not frequent):
- Electronics: Sold less often, massive revenue impact
```

In [47]:
%%sql
SELECT
  p.product_id,
  p.product_category,
  SUM(o.quantity) AS units_sold,
  SUM(o.total_amount) AS revenue
FROM orders o
JOIN products p ON o.product_id = p.product_id
GROUP BY p.product_id, p.product_category
ORDER BY units_sold DESC
LIMIT 20;


 * mysql+pymysql://root:***@localhost:3306
20 rows affected.


[(712, 'Books', Decimal('17'), Decimal('396.50')),
 (606, 'Food', Decimal('13'), Decimal('714.42')),
 (3238, 'Food', Decimal('13'), Decimal('273.78')),
 (7916, 'Toys', Decimal('13'), Decimal('1231.75')),
 (2391, 'Food', Decimal('13'), Decimal('520.78')),
 (4814, 'Beauty', Decimal('12'), Decimal('1101.98')),
 (2871, 'Books', Decimal('12'), Decimal('523.51')),
 (1495, 'Beauty', Decimal('11'), Decimal('622.93')),
 (11690, 'Books', Decimal('11'), Decimal('861.70')),
 (6029, 'Books', Decimal('11'), Decimal('252.00')),
 (2842, 'Books', Decimal('10'), Decimal('218.49')),
 (1244, 'Beauty', Decimal('10'), Decimal('257.66')),
 (549, 'Books', Decimal('10'), Decimal('222.14')),
 (2938, 'Sports', Decimal('10'), Decimal('3766.06')),
 (1453, 'Books', Decimal('10'), Decimal('837.80')),
 (3015, 'Books', Decimal('10'), Decimal('302.00')),
 (766, 'Books', Decimal('10'), Decimal('465.96')),
 (3000, 'Toys', Decimal('10'), Decimal('1153.55')),
 (2463, 'Beauty', Decimal('10'), Decimal('559.11')),
 (932, 'Books', Decimal('10'), Decimal('459.10'))]

<br>

## 11. Average Unit Price Per Category

**Highest:** Electronics ($1,767) | **Lowest:** Books ($57)

### 💡 Key Insights:
- **Massive Price Gap:** 31x difference between highest and lowest category
- **Clear Pricing Tiers:**
  - Premium: Electronics ($1,767), Home & Garden ($692)
  - Mid-tier: Sports ($495), Fashion ($278)
  - Budget: Toys ($170), Beauty ($114), Food ($72), Books ($57)

### 📊 Strategic Pricing Analysis:
```
HIGH MARGIN POTENTIAL:
Electronics: $1,767 avg → 20-30% margin = $350-530 per unit
Home & Garden: $692 avg → 25-35% margin = $173-242 per unit

LOW MARGIN REALITY:
Books: $57 avg → 5-15% margin = $3-9 per unit
Food: $72 avg → 10-20% margin = $7-14 per unit


In [48]:
%%sql
SELECT
  product_category,
  AVG(unit_price) AS avg_unit_price,
  COUNT(DISTINCT product_id) AS distinct_products
FROM products
GROUP BY product_category
ORDER BY avg_unit_price DESC;


 * mysql+pymysql://root:***@localhost:3306
8 rows affected.


[('Electronics', Decimal('1767.181478'), 2071),
 ('Home & Garden', Decimal('691.726130'), 2049),
 ('Sports', Decimal('494.867425'), 2214),
 ('Fashion', Decimal('278.341776'), 2016),
 ('Toys', Decimal('170.365698'), 2020),
 ('Beauty', Decimal('113.642746'), 2083),
 ('Food', Decimal('71.742085'), 1971),
 ('Books', Decimal('57.486800'), 1978)]

<br>

## 12. Which Categories Have the Highest Quantity Sold?

**Volume Leaders:** Sports (6,856) | Books (6,727) | Beauty (6,564)

### 💡 Key Insights:
- **Volume Champions ≠ Revenue Champions:**
  - Sports: #1 in units (6,856) but #3 in revenue ($3.21M)
  - Books: #2 in units (6,727) but #8 in revenue ($360K)
  - Electronics: #6 in units (6,277) but #1 in revenue ($10.48M)

- **Volume Distribution is Balanced:**
  - Range: 6,078 (Fashion) to 6,856 (Sports)
  - Only 13% variance across all categories
  - Suggests equal popularity but VERY different value

### 📊 Volume vs Revenue Disconnect:
```
HIGH VOLUME + HIGH REVENUE (Winners):
✅ Sports: 6,856 units = $3.21M (good unit economics)
✅ Electronics: 6,277 units = $10.48M (excellent unit economics)


In [49]:
%%sql
SELECT
  p.product_category,
  SUM(o.quantity) AS total_quantity
FROM orders o
JOIN products p ON o.product_id = p.product_id
GROUP BY p.product_category
ORDER BY total_quantity DESC;


 * mysql+pymysql://root:***@localhost:3306
8 rows affected.


[('Sports', Decimal('6856')),
 ('Books', Decimal('6727')),
 ('Beauty', Decimal('6564')),
 ('Food', Decimal('6381')),
 ('Toys', Decimal('6303')),
 ('Electronics', Decimal('6277')),
 ('Home & Garden', Decimal('6155')),
 ('Fashion', Decimal('6078'))]

<br>

## 13. Average Session Duration by Device

**Longest:** Tablet (14.61 min) | Desktop (14.53 min) | Mobile (14.53 min)

### 💡 Key Insights:
- **Minimal Variance:** Only 5 seconds difference across devices (essentially equal)
- **High Engagement:** 14.5 minutes is STRONG for e-commerce (industry avg: 5-7 min)
- **Pages Viewed:** ~9 pages average suggests good site exploration

### 📊 Performance Assessment:
```
Your Site: 14.5 min session, 9 pages
Industry Benchmark: 6-7 min session, 4-5 pages


In [53]:
%%sql
SELECT
  device_type,
  AVG(session_duration_minutes) AS avg_session_duration,
  AVG(pages_viewed) AS avg_pages
FROM customer_behavior
GROUP BY device_type
ORDER BY avg_session_duration DESC;


 * mysql+pymysql://root:***@localhost:3306
3 rows affected.


[('Tablet', Decimal('14.6105'), Decimal('9.1463')),
 ('Desktop', Decimal('14.5317'), Decimal('8.9596')),
 ('Mobile', Decimal('14.5250'), Decimal('9.0048'))]

<br>

## 14. How Many Pages Do Customers View on Average Before Purchasing?

**Result:** 9.0 pages per session

### 💡 Key Insights:
- **Industry Comparison:**
  - Your site: 9.0 pages
  - E-commerce benchmark: 4-6 pages
  - **You're 50-80% ABOVE industry average**


In [52]:
%%sql
SELECT AVG(pages_viewed) AS avg_pages_before_purchase FROM customer_behavior;


 * mysql+pymysql://root:***@localhost:3306
1 rows affected.


[(Decimal('9.0031'),)]

<br>

## 15. Which Device Types Lead to Higher Conversion-Rate Proxies?

**Volume Winner:** Mobile (9,543 orders - 56% of total)

### 💡 Key Insights:
- **Mobile Dominates Volume:**
  - Mobile: 9,543 orders (56%)
  - Desktop: 5,845 orders (34%)
  - Tablet: 1,661 orders (10%)

- **Desktop Has Higher AOV:**
  - Desktop: $1,311 average spend (+4% vs mobile)
  - Mobile: $1,260 average spend
  - Tablet: $1,258 average spend (lowest)

- **Mobile Opportunity:**
  - 56% of orders but 4% lower AOV = optimization opportunity
  - Bridging AOV gap: 9,543 orders × $51 = **$487K potential revenue gain**

### 📊 Device Strategy Matrix:
```
MOBILE (56% volume, $1,260 AOV):
Role: Volume Driver
Strength: Convenience, accessibility
Weakness: Lower spend per order
Customer: Impulse buyers, on-the-go shoppers

DESKTOP (34% volume, $1,311 AOV):
Role: Value Driver
Strength: Higher AOV, research-oriented
Weakness: Lower volume
Customer: Deliberate shoppers, B2B buyers

TABLET (10% volume, $1,258 AOV):
Role: Minimal Impact
Strength: None apparent
Weakness: Lowest volume AND lowest AOV
Customer: Casual browsers
`

In [54]:
%%sql
SELECT
  b.device_type,
  COUNT(DISTINCT b.order_id) AS orders_count,
  AVG(o.total_amount) AS avg_spend
FROM customer_behavior b
JOIN orders o ON b.order_id = o.order_id
GROUP BY b.device_type
ORDER BY orders_count DESC;


 * mysql+pymysql://root:***@localhost:3306
3 rows affected.


[('Mobile', 9543, Decimal('1260.472104')),
 ('Desktop', 5845, Decimal('1310.767677')),
 ('Tablet', 1661, Decimal('1257.634094'))]